### PR TITLE
[JUJU-3332] Support JWT login tokens for controller and model access

### PR DESCRIPTION
**Pull request by wallyworld**
Add support for JWT in the login request.
We support "controller" permissions (superuser, login) and "model" permissions (admin, write, read)

Use the controller config `login-token-refresh-url` to specify the endpoint to refresh the token public key from.
This can take 2 forms:
1.  `https://my.host`
2. `https://my.host/.wellknown/jwks.json`

In the first case, we append `/.wellknown/jwks.json`.

The token subject must be a user tag, like `user-fred`. ie who the request is being made on behalf of.

## QA steps

For now, just ensure that a bootstrap without `login-token-refresh-url` works as before.

The new functionality needs to be tested with JAAS, where the controller is started like so:

`juju bootstrap --config login-token-refresh-url=https://my.host ...`

### CODE REVIEW

File: apiserver/admin.go
Comment (ale8k) [inline apiserver/admin.go:410]: Would it be OK to move this struct into /pkg? So we can share the same struct?

File: apiserver/jwttoken.go
Comment (ale8k) [inline apiserver/jwttoken.go:86]: Can't we use token.Get("access")?

Comment (wallyworld) [reply apiserver/jwttoken.go:86]: We could but it would be a few extra LOC, because we'd need to check ok from that and then cast the value to a map and check that.

File: apiserver/admin.go
Comment (wallyworld) [reply apiserver/admin.go:410]: prefer not to because the all the juju over the wire params are kept together

File: apiserver/logintoken.go
Comment (hpidcock) [inline apiserver/logintoken.go:66]: `return nil, err`

All errors should return no token. More defensive. I've checked the libraries, it shouldn't return a token when an error is returned, but things could change.

File: apiserver/logintoken.go
Comment (hpidcock) [inline apiserver/logintoken.go:88]: It seems token.PrivateClaims() will always return a non-nil map, but it could be wise to check that.

File: apiserver/logintoken.go
Comment (hpidcock) [inline apiserver/logintoken.go:97]: we should call `.Validate()` on the `permission.Access`. I also noticed that `permission.Access.Validate` is missing a check for `permission.ConsumeAccess`.

File: apiserver/logintoken_test.go
Comment (hpidcock) [inline apiserver/logintoken_test.go:66]: Can we get a few tests that probe a few key places where the login should fail?

File: controller/config.go
Comment (hpidcock) [inline controller/config.go:1012]: They should also always use "https" as the scheme. Makes development more difficult, but here its important.

File: apiserver/apiserver.go
Comment (hpidcock) [inline apiserver/apiserver.go:371]: Does this refresh happen elsewhere?

**Review by hpidcock — APPROVED** [review]
Looks great, a few questions and concerns. Needs probably more testing, especially around failure scenarios.

**Comment by kian99** [comment]
Thanks for the work on this! After testing this I can confirm the JWT parsing all seems okay. Some comments on other items:

1. Will Juju still validate the user? When testing and running commands `juju models` I would get a `user "fred" not found` error. I believe previously when using JASS the Juju controller would need to be bootstrapped to point to Candid but now that JIMM is authenticating the users will Juju still need to do that?

2. The JIMM Rebac spec has a section as below
>If the user does not have any of the required access levels, JIMM will return a “permission denied” error to the client. JIMM will keep retrying the method call if it keeps returning the PermissionAssertionRequiredError with additional entities, adding additional claims to the JWT. 

Where do we create a list of the missing access rights and return that specific error? Or will this be a follow up PR?

**Comment by wallyworld** [comment]
> Thanks for the work on this! After testing this I can confirm the JWT parsing all seems okay. Some comments on other items:
> 
>     1. Will Juju still validate the user? When testing and running commands `juju models` I would get a `user "fred" not found` error. I believe previously when using JASS the Juju controller would need to be bootstrapped to point to Candid but now that JIMM is authenticating the users will Juju still need to do that?
> 

Juju will not maintain a local set of users when a login token is used. The controller will need to reject add-user commands etc. The controller will also not point to candid. The whole point of using a jwt is that an external source is managing the entire user and permissions model. There will be some fallout which will need to be fixed, where the controller tries to look up a local user.


>     2. The JIMM Rebac spec has a section as below
> 
> 
> > If the user does not have any of the required access levels, JIMM will return a “permission denied” error to the client. JIMM will keep retrying the method call if it keeps returning the PermissionAssertionRequiredError with additional entities, adding additional claims to the JWT.
> 
> Where do we create a list of the missing access rights and return that specific error? Or will this be a follow up PR?

That's not in this PR. This PR is just for controller and model permissions.

File: apiserver/apiserver.go
Comment (wallyworld) [reply apiserver/apiserver.go:371]: yes, this just seeds the initial key

File: apiserver/admin.go
Comment (wallyworld) [reply apiserver/admin.go:407]: We're passing in a known valid "kind", eg "controller", so the error can never happen. I'll add a comment.

File: apiserver/logintoken.go
Comment (wallyworld) [reply apiserver/logintoken.go:66]: jwt is nil when there's an error but i guess best to be explicit

File: apiserver/apiserver.go
Comment (wallyworld) [reply apiserver/apiserver.go]: redirects now disabled, https for the url enforced in config validation but no way to pass in a custom CA cert yet

Comment (wallyworld) [reply apiserver/apiserver.go]: I'll defer the check for https until config is updated to allow a certificate to be configured. This will unblock testing.

Comment (kian99) [reply apiserver/apiserver.go]: Btw for testing with https I push a custom CA cert into the lxc container running the controller and restart the container, then https works just fine without configuring the cert in the controller config directly. You can see how this is done in the instructions for the mitm service.

In case you want to add this now, we shouldn't be blocked.

**Pull request merged**
